### DISCUSSION TITLE
Turn Switch1 ON only after Switch2 turned OFF

### DISCUSSION THREAD

**Discussion by JanaufGitHub** [arendst/Tasmota#22047] [General]
**Question:** 
How can I send a request to Switch1 that will first reliably turn Switch2 off before turning Switch1 on?

**Context:** 
For my rudimentary home PV battery charging installation I would like to switch on the battery charger when my net export to the grid exceeds a certain limit (i.e. when I have enough surplus power I would otherwise loose). When the surplus falls below a certain limit I want to switch the charger off again. At other times (e.g. during the night) I want to activate a power inverter supplying battery electricity to my home grid. Obviously, I do not want to charge the battery while it feeds power to the grid via the inverter. So I want to make really sure that the connection of the inverter to the home grid is off when the charger is switched on. And when the inverter gets connected, the charger shall be switched off before.

**Solution Proposal:** 
I was contemplating using two Tasmota plug switches. One for the charger, the other for the inverter. When my RaspberryPi detects an overall export of more than the configured threshold, I could send an HTTP request to Switch2 (inverter) to turn it off and afterwards send a second request to Switch1 (charger) to turn it on. But the script handling this would need to wait until the Off status of Switch2 was confirmed and only then turn Switch1 on. This would block the script for some (timeout) time. So maybe placing this within the Tasmota switches would be better.

Many thanks for your help! 
Jan

**Comment by barbudor** [comment]
> But the script handling this would need to wait until the Off status of Switch2 was confirmed

Why ?
The switch change is immediate
On return of the call, it is already performed

**Reply by JanaufGitHub** [reply]
Indeed, if the call returns a code 0 or something to indicate correct execution, fine.
But if e.g. Switch2 could not be reached one could not turn Switch1 ON. Switch2 might still be running and this should not happen.
Does Tasmota return a code upon successfull execution of an http request?

**Reply by sfromis** [reply]
The command response could indicate if communication failed or not, but not the response without a custom build. Using `http` is not great if you want reliable communication, Tasmota is designed for MQTT.

Such details are why I mentioned the technique of having one device forward a command to the other, after having done the first step, but that would not "confirm" if the relay worked, or whatever. Hence, the word "confirm" is still pretty unclear about how little is "enough" for that, or if you want some sort of "verification" as confirmation.

**Comment by sfromis** [comment]
You could do something like sending a combined to command to one plug to switch power off, and then have it send a command to the other device to power on. But with no explanation of what "confirmed" means in operational terms, and what pause may be acceptable if slowing things down, too few details for specific suggestions.

**Reply by JanaufGitHub** [reply]
Hmm, yes, I suppose one could do it "backwards": 
In order to turn Switch1 ON send a command series to Switch2, turning it OFF and afterwards turn Switch1 on from by having Switch2 send the ON command.

Do you guys know:
If you send a backlog of commands to a Tasmota device, and one command in the backlog can not be executed (e.g. the http request because the device to which it was sent is offline), does the rest of the backlog still execute - or is it aborted?

**Reply by sfromis** [reply]
No abort will happen. Again, the question is how seriously you mean it when wanting "confirmation".

You could also get a switch where the same relay will switch one terminal between two on the other side.

**Comment by JanaufGitHub** [comment]
OK, let's maybe not get hung up on the expression "confirmation". I don't need an audit trail or so.
I just really want to make sure I only charge the battery when connection of the inverter to the house grid is off,
and I only start injecting power into the house grid when the battery's charger is off.
And this should be resilient enough that I don't end up feeding power from the battery back into the battery plus some extra to account for the losses... ;-)

**Reply by sfromis** [reply]
I already suggested a simple and resilient solution, a single-pole double-throw relay, able to connect one wire to either of two on the other side. As it is physically moving the connecting part, it would have be very broken to not do what you want, without hassle or need for planning for contingencies. There are many such in the "Relay Board" section on https://templates.blakadder.com/relay.html

While such a relay would not allow you to have both off; if you want that too, you could get one with two relays, and wire one relay for switching the output on/off, and the other for switching between two inputs. Note that such boards also needs a power source, as they have dry relays, and if you want to have backup power either way, you cannot be sure which, if any, input will always provide power. That complicates things again :grinning:

**Reply by JanaufGitHub** [reply]
Hi @sfromis,

hmm, yes. One could also solve this via appropriate hardware.
But you are right, a single-pole double-throw relay would always have one ON and the other off. 
Not sure if this would work for me to always drain the battery (Switch2 ON) when it can't be charged (Switch1 OFF).
There are pros and cons.

**Thanks y'all!**
